### ISSUE TITLE
Avoid invocation of subsequent AuthorizationHandlers for a single Requirement

### ISSUE THREAD

**Issue by marwalsch** [dotnet/aspnetcore#36536]
Consider the following `Requirement` with two resource-based handlers:

```
public class HasPermissionHandler1: AuthorizationHandler<HasPermissionRequirement, string>
    {
        protected override Task HandleRequirementAsync(AuthorizationHandlerContext context, HasPermissionRequirement requirement, string permissionName)
        {
            // checks one way a permission can be granted and then succeeds
        }
    }

    public class HasPermissionHandler2 : AuthorizationHandler<HasPermissionRequirement, string>
    {
        protected override Task HandleRequirementAsync(AuthorizationHandlerContext context, HasPermissionRequirement requirement, string permissionName)
        {
            // checks another way a permission can be granted and then succeeds
        }
    }
```
There a different ways in which a permission can be assigned to a user, but one suffices to pass the `Requirement`.

However, i noticed that regardless of whether each or any of the handlers succeed, the `HandleRequirementAsync()` method is always invoked for each handler. I wonder why this is the case. The handlers act disjunctively, which is why a call to `HasPermissionHandler2 `would be unnecessary if `HasPermissionHandler1` already returned a successful result.
This is of interest to me primarily due to performance reasons. For instance, if there is a handler checking for in-memory claims granting permissions, I don't want additional handlers to hit the database to determine permissions through other means (i.e. roles, groups,...).
Obviously I could just define a single handler that contains a concatenation of all these checks, but the given one appears a as a much cleaner solution.
(This performance concern also involves the reliability of the order the different handlers are invoked. It appears that it is equal 
to the order the services are registered; is this reliable?)

To sum up:

Current behaviour:
- `HandleRequirementAsync()` is invoked for each handler of the respective requirement.

Desired behaviour:
- Subsequent calls to `HandleRequirementAsync()` are omitted for different handlers as soon as one succeeds.

**Comment by abdusco** [comment]
The reason all handlers are called is that one handler can explicitly **fail** a requirement, even if some other handler has already marked it as **succeeded**. That's why all handlers must be executed to figure out if the requirement is actually fulfilled.

From [the docs](https://docs.microsoft.com/en-us/aspnet/core/security/authorization/policies?view=aspnetcore-5.0#what-should-a-handler-return):
> If a handler calls context.Succeed or context.Fail, all other handlers are still called. This allows requirements to produce side effects, such as logging, which takes place even if another handler has successfully validated or failed a requirement. When set to false, the InvokeHandlersAfterFailure property short-circuits the execution of handlers when context.Fail is called. InvokeHandlersAfterFailure defaults to true, in which case all handlers are called.

**Comment by marwalsch** [comment]
@abdusco You are right, I didn't keep `context.Fail ` in mind as I never came across a scenario where it was reasonable to let a requirement fail explicitly; This also appears to contradict the idea of disjunctively connecting handlers in the way shown above; If you have to invoke every handler anyway, why fail explicitly if the handlers should just not succeed anyway; every policy that can explicitly fail in such case should be separate one to begin with: and just not succeed on the inverse check (at least from my personal experience).
Considering, how about a flag that could be set when registering the policy to explicitly tell the middleware that one successful handler results in overall success. 

```
enum HandlerResolutionMode {Strict, Lax}

services.AddAuthorization(options =>
{
    options.AddPolicy(Policies.HasPermission, policy => policy.Requirements.Add(new HasPermissionRequirement(), HandlerResolutionMode.Lax));
});
```

**Comment by blowdart** [comment]
As @abdusco mentions this is by design and won't change. Handlers can have side effects like logging, and we deliberately run through them all, even if one passes. You're also assuming they're called in a particular order which we don't guarantee.

**Comment by marwalsch** [comment]
@blowdart That's a pity but sounds reasonable. Any recommendation on achieving a true disjunction other than decorating the handlers and breaking out of the checks manually if one succeeds?

**Comment by marwalsch** [comment]
As of now I am writing every handler that is checked for on my own. Regarding the `IsFired`-Policy my approach based on my understanding around policies would be to write a handler that asserts whether the current user is employed (`!= fired`) and check upon it respectively; in fact, I already have a quite similar mechanism regarding inactive (=suspended, discharged, comatose,...) employees. It works well for these crucial things when setting them as a default policy.

Speaking of examples: The `OfficeEntryRequirement` from the docs is fulfilled through possession of a badge _or_ a temporary stamp represented through two respective handlers. As expected either one of these suffices for access, but the idea of running all handlers because one may subsequently fail is actually very counter-intuitive. If you have a temporary stamp no one would double check the badge as well to make sure the check will pass (or `fail`, for instance when the badge is a forgery; which, again, I would just treat through omitting `context.succeed`). – Merely to demonstrate where my confusion comes from.

I will have to elaborate whether the subsequent policy checks are as much of an issue as I fear them to be, but I assume my worries are understandable:
Let the two handlers in the example above hit the database and consider an additional `HasPermissionHandler0` that checks for permissions through claims rather than the database. It is invoked through a Blazor Server page with 10 `<AuthorizeView/>` components.
With prioritizing the claim checking handler and omitting unnecessary calls, one might actually fulfil all policies without even hitting the database once, which would only be necessary for permissions not given (two round trips to determine it can not be fulfilled in `HasPermissionHandler1` nor `HasPermissionHandler2`). However, in the given scenario, it is 20 round-trips every time guaranteed.

**Comment by blowdart** [comment]
Oh certainly your concerns are understandable, I'll until you realise why we run through everything it can be counter intuitive. But it's safer/more secure, that's the default path we want to have and why the docs explain that we run every requirement handler.

**Comment by marwalsch** [comment]
I understand, I will reconsider my approach. Thanks for the input.
